src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs

```diff
@@ -0,0 +1,129 @@
+    {
+        public static dynamic DynamicEvents(this TraceGC traceGC)
+        {
+            return new DynamicIndex(traceGC.dynamicEventIndex);
+        }
+    }
+
+    // TODO, AndrewAu, this could be compiled for better performance, right now we are doing linear lookups
+    // We can do validation so that the input make sense (i.e. no unexpected types, no duplicate event/field), etc
+    //
+    public class DynamicEventSchema
+    {
+        internal static List<DynamicEventSchema> DynamicEventSchemas = new List<DynamicEventSchema>();
+
+        public string DynamicEventName { get; set; }
+
+        public List<KeyValuePair<string, Type>> Fields { get; set; }
+
+        public static void Set(List<DynamicEventSchema> dynamicEventSchemas)
+        {
+            DynamicEventSchemas = dynamicEventSchemas;
+        }
+    }
+
+    internal class DynamicIndex : DynamicObject
+    {
+        private Dictionary<string, List<RawDynamic>> index;
+
+        public DynamicIndex(Dictionary<string, List<RawDynamic>> index)
+        {
+            this.index = index;
+        }
+
+        public override bool TryGetMember(GetMemberBinder binder, out object result)
+        {
+            if (DynamicEventSchema.DynamicEventSchemas.Any(s => string.Equals(s.DynamicEventName, binder.Name)))
+            {
+                string name = binder.Name;
+                if (index.TryGetValue(name, out var indexValue))
+                {
+                    // TODO, andrewau, handle the case where we have multiple dynamic event of the same name in a single GC.
+                    //
+                    // That should be an array of these dynamic events
+                    // Somewhere the schema should tell us if this is expected or not so not to generate wrong stuff just because the data is corrupted.
+                    // 
+                    result = new DynamicEvent(indexValue[0], DynamicEventSchema.DynamicEventSchemas.Single(s => string.Equals(s.DynamicEventName, binder.Name)).Fields);
+                    return true;
+                }
+                else
+                {
+                    result = null;
+                    return true;
+                }
+            }
+            else
+            {
+                result = null;
+                return false;
+            }
+        }
+    }
+
+    internal class DynamicEvent : DynamicObject
+    {
+        private RawDynamic rawDynamic;
+        private Dictionary<string, object> fieldValues;
+
+        public DynamicEvent(RawDynamic rawDynamic, List<KeyValuePair<string, Type>> fields)
+        {
+            this.rawDynamic = rawDynamic;
+            this.fieldValues = new Dictionary<string, object>();
+            int offset = 0;
+            foreach (KeyValuePair<string, Type> field in fields)
+            {
+                object value = null;
+                // TODO, AndrewAu, all types that we envision this will be needed
+                if (field.Value == typeof(ushort))
```

Comment:
What about arrays of numerics? Should we handle those in a first pass?

Reply:
I don't think we can emit array of numbers right now in dynamic event, can we?